src/middleware.js

```diff
@@ -18,50 +18,51 @@ const registrar = list => fn => {
 
 const createMiddleware = extraArgument => {
   const subscribers = []
-  const executors = {}
   let queuePromise = Promise.resolve()
 
   const middleware = store => next => action => {
-    const executeCommand = cmd => {
-      const result = try_(() => executors[cmd.type](cmd.command))
+    const executeEffect = effect => {
+      const result = try_(() => effect.run())
       return result.fold(
         error => {
-          const downstreamPromise = cmd.actionType
-            ? store.dispatch(fromError(cmd.actionType, error))
+          const downstreamPromise = effect.resultActionType
+            ? store.dispatch(fromError(effect.resultActionType, error))
             : Promise.resolve()
           return [result, downstreamPromise]
         },
         value => {
           if (isPromise(value)) {
             const downstreamPromise = value.then(
               resolvedValue => {
-                return cmd.actionType
-                  ? store.dispatch(fromSuccess(cmd.actionType, resolvedValue))
+                return effect.resultActionType
+                  ? store.dispatch(
+                      fromSuccess(effect.resultActionType, resolvedValue)
+                    )
                   : Promise.resolve()
               },
               error => {
-                if (cmd.actionType) {
+                if (effect.resultActionType) {
                   store
-                    .dispatch(fromError(cmd.actionType, error))
+                    .dispatch(fromError(effect.resultActionType, error))
                     .catch(() => {})
                 }
                 throw error
               }
             )
             return [result, downstreamPromise]
           } else {
-            const downstreamPromise = cmd.actionType
-              ? store.dispatch(fromSuccess(cmd.actionType, value))
+            const downstreamPromise = effect.resultActionType
+              ? store.dispatch(fromSuccess(effect.resultActionType, value))
               : Promise.resolve()
             return [result, downstreamPromise]
           }
         }
       )
     }
 
-    const runImmediateCommands = commands => {
-      const immediateCommands = commands.filter(isImmediateCommand)
-      const promiseTuples = immediateCommands.map(executeCommand)
+    const runImmediateEffects = effects => {
```

Comment:
What about effect structure validation? Make sure it has `run` and `isQueued`?

Reply:
non valid effects are filtered in line #150.
do you think we should throw an exception / show warning in case of non production env ?

Reply:
We can do structure validation in dev environments only.

`isQueued` should test against `true/false` and not truthy/falsy. While if there's no run, nothing will run in production.

Reply:
Effectively, this is what the helpers in `effects.js` are doing.